pmd-core/src/main/java/net/sourceforge/pmd/properties/PropertyBuilder.java

```diff
@@ -0,0 +1,370 @@
+     * and invalid values will be reported. A rule will never be run
+     * if some of its properties violate some constraints.
+     *
+     * @param constraint The constraint
+     *
+     * @return The same builder
+     */
+    // TODO we could probably specify the order of execution of constraints come 7.0.0, for now this remains unspecified
+    @SuppressWarnings("unchecked")
+    public B require(PropertyConstraint<? super T> constraint) {
+        validators.add(constraint);
+        return (B) this;
+    }
+
+
+    /**
+     * Specify a default value.
+     *
+     * @param val Value
+     *
+     * @return The same builder
+     */
+    @SuppressWarnings("unchecked")
+    public B defaultValue(T val) {
+        this.defaultValue = val;
+        return (B) this;
+    }
+
+
+    /**
+     * Builds the descriptor and returns it.
+     *
+     * @return The built descriptor
+     *
+     * @throws IllegalStateException if the default value doesn't satisfy the given constraints
+     */
+    public abstract PropertyDescriptor<T> build();
+
+
+    /**
+     * Returns the name of the property to be built.
+     */
+    public String getName() {
+        return name;
+    }
+
+
+    /**
+     * Generic builder for a single-value property.
+     *
+     * @param <T> Type of values the property handles
+     *
+     * @author Clément Fournier
+     * @since 6.10.0
+     */
+    // Note: we may keep some specialized property builders around to allow for some sugar,
+    // e.g. specifying the default value of a regex property as a string, or like the collection one,
+    // with varargs for collection types
+    public static final class GenericPropertyBuilder<T> extends PropertyBuilder<GenericPropertyBuilder<T>, T> {
+
+
+        private final ValueParser<T> parser;
+        private final Class<T> type;
+
+
+        GenericPropertyBuilder(String name, ValueParser<T> parser, Class<T> type) {
+            super(name);
+            this.parser = parser;
+            this.type = type;
+        }
+
+
+        protected ValueParser<T> getParser() {
+            return parser;
+        }
+
+
+        protected Class<T> getType() {
+            return type;
+        }
+
+
+        /**
+         * Returns a new builder that can be used to build a property
+         * handling lists of Ts. The validators already added are
+         * converted to list validators. The default value cannot have
+         * previously been set.
+         *
+         * @return A new list property builder
+         *
+         * @throws IllegalStateException if the default value has already been set
+         */
+        /* package private */ GenericCollectionPropertyBuilder<T, List<T>> toList() {
+
+            Supplier<List<T>> listSupplier = new Supplier<List<T>>() {
+                @Override
+                public List<T> get() {
+                    return new ArrayList<>();
+                }
+            };
```

Comment:
do we really need this and to pass it around just to create an empty list? sounds like overkill to me

Reply:
This is in prevision of the migration to Java 8. Many things here are sloppy because I designed this as if it was Java 8, but had to write it with Java 7 syntax so it's ugly. In Java 8 it will be as short as `ArrayList::new`, and we can also provide a `toSet` method with `HashSet::new`, or let the user use any collection supplier they want.

Reply:
I'm unsure there is a need for arbitrary collections types at all, and I fear being flexible about it is overengineering. Do we even have a use case for this?

Reply:
It's extremely easy to grab, so I wouldn't consider that overengineering. Everything is already completely generic, and GenericMultivaluePropertyDescriptor will be removed with 7.0.0, so there's really no point in artificially constraining ourselves to lists. One use case I can think of is the options for metrics rules, which would be better served by using a Set rather than a list.